ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java

```diff
@@ -0,0 +1,151 @@
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.json.JSONObject;
+
+
+/**
+ * Handles S3 requests for listing blobs that start with a provided prefix.
+ * API reference: <a href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html">...</a>
+ */
+public class S3ListHandler {
+  private final NamedBlobListHandler namedBlobListHandler;
+  private RestRequest restRequest;
+  private static final String PREFIX_PARAM_NAME = "prefix";
+  private static final String MAXKEYS_PARAM_NAME = "max-keys";
+  private static final String DELIMITER_PARAM_NAME = "delimiter";
+  private static final String ENCODING_TYPE_PARAM_NAME = "encoding-type";
+  private static final Charset CHARSET = StandardCharsets.UTF_8;
+  private static final ObjectMapper objectMapper = new XmlMapper();
+
+  /**
+   * Constructs a handler for handling s3 requests for listing blobs.
+   * @param namedBlobListHandler named blob list handler
+   */
+  public S3ListHandler(NamedBlobListHandler namedBlobListHandler) {
+    this.namedBlobListHandler = namedBlobListHandler;
+  }
+
+  /**
+   * Asynchronously get account metadata.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   */
+  public void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> callback) {
+    this.restRequest = restRequest;
+    namedBlobListHandler.handle(restRequest, restResponseChannel, ((result, exception) -> {
+      if (exception != null) {
+        callback.onCompletion(result, exception);
+        return;
+      }
+
+      try {
+        // Convert from json response to S3 xml response as defined in
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_ResponseSyntax.
+        RetainingAsyncWritableChannel writableChannel = new RetainingAsyncWritableChannel();
+        result.readInto(writableChannel, null).get();
+        JSONObject jsonObject = FrontendUtils.readJsonFromChannel(writableChannel);
+        Page<NamedBlobListEntry> page = Page.fromJson(jsonObject, NamedBlobListEntry::new);
+        ReadableStreamChannel readableStreamChannel = serializeAsXml(page);
+        restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/xml");
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, readableStreamChannel.getSize());
+        callback.onCompletion(readableStreamChannel, null);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
```

Comment:
This should be a 503 status. Let's just return a runtime exception here.

Reply:
Done..